### PR TITLE
fix member injection for supers from a different module

### LINKED ISSUE #438: [Generated Factories] Another MembersInjectors for super classes issue?
This looks an awful lot like https://github.com/square/anvil/issues/343

I confirmed this issue still exists when I update to latest Dagger and Anvil.  (2.40.5 and 2.3.10)
```
GradleProject1
--------------
class A : B <-- Well-formed MembersInjector
class B : C <-- broken

GradleProject2
--------------  
class C <-- Well-formed MembersInjector
```

When I inserted an intermediary class into the chain it made it clear that it breaks at the boundary between gradle projects.
```
GradleProject1
--------------
class A : B <-- Well-formed MembersInjector
class B : C <-- broken

GradleProject2
--------------  
class C : D <-- Well-formed MembersInjector
class D <-- Well-formed MembersInjector
```
```
GradleProject1
--------------
class A : B <-- Well-formed MembersInjector
class B : C <-- Well-formed MembersInjector
class C : D <-- broken

GradleProject2
--------------  
class D <-- Well-formed MembersInjector
```

Do I need to make a sample project, or is this good enough?

### LINKED ISSUE #439: constructor inject factory does not include member injection if direct super is in a different module
from a comment by @gmarques33  here: https://github.com/square/anvil/issues/438#issuecomment-988356976

I just ran into a pretty similar issue, but as I was using constructor injection it was a `Factory` instead of a `MembersInjector`.

In my case I have:

// :lib1
```kotlin
abstract class Base {
  @Inject lateinit var dep1: Dep1
}
```

// :lib2
```kotlin 
class Impl @Inject constructor(private val dep2: Dep2) : Base() {
}
```

On my DaggerApplicationComponent it will correctly generate:
```java
this.ImplProvider = Impl_Factory.create(dep2Provider, dep1Provider);
```

But the `Impl_Factory` generated by Anvil will only have a `create(dep2: Provider<Dep2>)` method:
```kotlin
import com.myproject.Dep2
import dagger.`internal`.Factory
import javax.inject.Provider
import kotlin.Suppress
import kotlin.jvm.JvmStatic

public class Impl_Factory(
  private val dep2: Provider<Dep2>
) : Factory<Impl> {
  public override fun `get`(): Impl = newInstance(dep2.get())

  public companion object {
    @JvmStatic
    public fun create(dep2: Provider<Dep2>):
        Impl_Factory = Impl_Factory(dep2)

    @JvmStatic
    public fun newInstance(dep2: Dep2): Impl =
        Impl(dep2)
  }
}
```

If I use kapt, Dagger will generate:
```Java
// Generated by Dagger (https://dagger.dev).
import com.myproject.Base_MembersInjector;
import com.myproject.Dep1;
import dagger.internal.DaggerGenerated;
import dagger.internal.Factory;
import javax.inject.Provider;

@DaggerGenerated
@SuppressWarnings({
    "unchecked",
    "rawtypes"
})
public final class Impl_Factory implements Factory<Impl> {
  private final Provider<Dep2> dep2Provider;

  private final Provider<Dep1> dep1Provider;

  public Impl_Factory(Provider<Dep2> dep2Provider,
      Provider<Dep1> dep1Provider) {
    this.dep2Provider = dep2Provider;
    this.dep1Provider = dep1Provider;
  }

  @Override
  public Impl get() {
    Impl instance = newInstance(dep2Provider.get());
    Base_MembersInjector.injectDep1(instance, dep1Provider.get());
    return instance;
  }

  public static Impl_Factory create(
      Provider<Dep2> dep2Provider,
      Provider<Dep1> dep1Provider) {
    return new Impl_Factory(dep2Provider, dep1Provider);
  }

  public static Impl newInstance(Dep2 dep2) {
    return new Impl(dep2);
  }
}
```

_Originally posted by @gmarques33 in https://github.com/square/anvil/issues/438#issuecomment-988356976_

### PR DESCRIPTION
**Pull request by RBusarow**
fixes #438
fixes #439

### CODE REVIEW

File: compiler/src/test/java/com/squareup/anvil/compiler/dagger/InjectConstructorFactoryGeneratorTest.kt [hunk 1]
```diff
@@ -1049,6 +1049,167 @@ public class InjectClass_Factory(
...
+      import javax.inject.Provider
+      import dagger.Lazy
+
+      class InjectClass @Inject constructor() : Base() {
+
+        @Inject
+        lateinit var name: String
+      }
+      """,
+      previousCompilationResult = otherModuleResult
```
Comment (RBusarow) [inline compiler/src/test/java/com/squareup/anvil/compiler/dagger/InjectConstructorFactoryGeneratorTest.kt:1086]: Again, @vRallev, this was such a great idea!

Chaining the compilations makes it so much easier to write these tests without creating one-off test classes elsewhere.

**Review by gmarques33 — COMMENTED** [review]
Thanks for working on it so quickly @RBusarow

File: compiler/src/test/java/com/squareup/anvil/compiler/dagger/InjectConstructorFactoryGeneratorTest.kt [hunk 2]
```diff
@@ -1049,6 +1049,167 @@ public class InjectClass_Factory(
...
+      abstract class Base {
+
+        @Inject
+        lateinit var base1: List<Int>
+
+        @Inject
+        lateinit var base2: List<String>
+      }
+      """
+    )
```
Comment (vRallev) [inline compiler/src/test/java/com/squareup/anvil/compiler/dagger/InjectConstructorFactoryGeneratorTest.kt:1069]: I think it's good practice here to verify that the compilation result is successful nonetheless:
```kotlin
) {
  assertThat(result).isEqualTo(OK)
}
```

File: compiler/src/test/java/com/squareup/anvil/compiler/dagger/InjectConstructorFactoryGeneratorTest.kt [hunk 3]
```diff
@@ -1049,6 +1049,167 @@ public class InjectClass_Factory(
...
```
Comment (vRallev) [inline compiler/src/test/java/com/squareup/anvil/compiler/dagger/InjectConstructorFactoryGeneratorTest.kt:1141]: Same here.

**Review by vRallev — APPROVED** [review]

**Review by vRallev — APPROVED** [review]

**Comment by vRallev** [comment]
Thanks Rick!

**Pull request merged**